src/main/java/dan200/computercraft/client/gui/widgets/WidgetTerminal.java

```diff
@@ -221,37 +226,64 @@ public void handleMouseInput( int mouseX, int mouseY )
                 charX = Math.min( Math.max( charX, 0 ), term.getWidth() - 1 );
                 charY = Math.min( Math.max( charY, 0 ), term.getHeight() - 1 );
 
-                if( m_lastClickButton >= 0 && !Mouse.isButtonDown( m_lastClickButton ) )
-                {
-                    if( m_focus ) computer.mouseUp( m_lastClickButton + 1, charX + 1, charY + 1 );
-                    m_lastClickButton = -1;
-                }
+                handleMouseClick( computer, charX, charY );
+                handleMouseWheel( computer, charX, charY );
+                handleMouseMove( computer, charX, charY );
+            }
+        }
+        else // The mouse has moved out of the terminal, send a -1, -1 mouse_move event
+        {
+            handleMouseMove( computer, -1, -1 );
+        }
+    }
 
-                int wheelChange = Mouse.getEventDWheel();
-                if( wheelChange == 0 && m_lastClickButton == -1 )
-                {
-                    return;
-                }
+    private void handleMouseClick( IComputer computer, int charX, int charY )
+    {
+        if( m_lastClickButton >= 0 && !Mouse.isButtonDown( m_lastClickButton ) )
+        {
+            if( m_focus ) computer.mouseUp( m_lastClickButton + 1, charX + 1, charY + 1 );
+            m_lastClickButton = -1;
+        }
+    }
 
-                if( m_focus )
-                {
-                    if( wheelChange < 0 )
-                    {
-                        computer.mouseScroll( 1, charX + 1, charY + 1 );
-                    }
-                    else if( wheelChange > 0 )
-                    {
-                        computer.mouseScroll( -1, charX + 1, charY + 1 );
-                    }
+    private void handleMouseWheel( IComputer computer, int charX, int charY )
+    {
+        int wheelChange = Mouse.getEventDWheel();
+        if( wheelChange == 0 && m_lastClickButton == -1 )
```

Comment:
Instead of moving these into separate functions, I'd keep it as is and just delete this if block. Yes, `handleMouseInput` will be an ugly behemoth, but it is already.